File: src/handlers/webhook.py
import json
import os
import boto3
from linebot import LineBotApi, WebhookHandler
from linebot.models import MessageEvent, TextMessage, TextSendMessage
import uuid
from datetime import datetime

line_bot_api = LineBotApi(os.environ['LINE_CHANNEL_ACCESS_TOKEN'])
handler = WebhookHandler(os.environ['LINE_CHANNEL_SECRET'])
dynamodb = boto3.resource('dynamodb')
table = dynamodb.Table('Reminders')

def lambda_handler(event, context):
    signature = event['headers'].get('X-Line-Signature', '')
    body = event['body']
    
    @handler.add(MessageEvent, message=TextMessage)
    def handle_message(line_event):
        text = line_event.message.text
        if text.startswith('!remind'):
            result = parse_reminder_command(text)
            if result:
                schedule_pattern, reminder_time, message = result
                dynamodb_util = DynamoDBUtil('Reminders')
                dynamodb_util.create_reminder(
                    line_event.source.group_id,
                    message,
                    reminder_time,
                    schedule_pattern
                )
                line_bot_api.reply_message(
                    line_event.reply_token,
                    TextSendMessage(text=f'リマインダーを設定しました: {message} ({reminder_time})')
                )
        elif text.startswith('!list'):
            list_reminders(line_event)

    try:
        handler.handle(body, signature)
    except Exception as e:
        return {
            'statusCode': 400,
            'body': json.dumps({'error': str(e)})
        }

    return {
        'statusCode': 200,
        'body': json.dumps({'message': 'OK'})
    }

def parse_reminder_command(text):
    """
    !remind monthly 15 10:00 会議の時間です
    のような形式をパース
    """
    parts = text.split()
    if len(parts) < 5:
        return None
    
    if parts[1] == 'monthly':
        try:
            day = int(parts[2])
            time = parts[3]
            message = ' '.join(parts[4:])
            if 1 <= day <= 31 and len(time.split(':')) == 2:
                schedule_pattern = f"monthly_{day}_{time}"
                next_reminder = get_next_monthly_reminder(day, time)
                return schedule_pattern, next_reminder, message
        except ValueError:
            pass
    return None

def get_next_monthly_reminder(day, time):
    now = datetime.now()
    if now.day > day or (now.day == day and now.strftime('%H:%M') > time):
        # 来月の指定日
        next_month = now.month + 1 if now.month < 12 else 1
        next_year = now.year if now.month < 12 else now.year + 1
    else:
        # 今月の指定日
        next_month = now.month
        next_year = now.year
    
    return f"{next_year}-{next_month:02d}-{day:02d} {time}"

File: src/handlers/webhook.ts
import { APIGatewayProxyHandler } from 'aws-lambda';
import { Client, validateSignature, WebhookEvent } from '@line/bot-sdk';
import { DynamoService } from '../services/dynamo-service';

const config = {
    channelAccessToken: process.env.LINE_CHANNEL_ACCESS_TOKEN!,
    channelSecret: process.env.LINE_CHANNEL_SECRET!
};

const lineClient = new Client(config);
const dynamoService = new DynamoService();

export const handler: APIGatewayProxyHandler = async (event) => {
    if (!validateSignature(event.body!, config.channelSecret, event.headers['x-line-signature'])) {
        return { statusCode: 403, body: 'Invalid signature' };
    }

    const body = JSON.parse(event.body!);
    const events: WebhookEvent[] = body.events;

    for (const lineEvent of events) {
        if (lineEvent.type === 'message' && lineEvent.message.type === 'text') {
            const { text } = lineEvent.message;
            
            if (text === '!remind list') {
                // リマインド一覧表示
                const reminders = await dynamoService.listReminders(lineEvent.source.groupId!);
                const reminderList = reminders.map(r => {
                    if (r.isRecurring && r.recurringPattern) {
                        return `ID: ${r.id} (毎月${r.recurringPattern.dayOfMonth}日${r.recurringPattern.hour}:${r.recurringPattern.minute}) ${r.message}`;
                    }
                    return `ID: ${r.id} (${new Date(r.scheduleTime).toLocaleString()}) ${r.message}`;
                }).join('\n');
                
                await lineClient.replyMessage(lineEvent.replyToken, {
                    type: 'text',
                    text: reminderList || 'リマインドはありません'
                });
            } else if (text.startsWith('!remind delete ')) {
                // リマインド削除
                const id = text.split(' ')[2];
                await dynamoService.deleteReminder(id);
                await lineClient.replyMessage(lineEvent.replyToken, {
                    type: 'text',
                    text: `ID: ${id} のリマインドを削除しました`
                });
            } else if (text.startsWith('!remind ')) {
                const parts = text.split(' ');
                const timeSpec = parts[1];
                const message = parts.slice(2).join(' ');

                // 毎月指定の形式: "毎月15日10:30"
                if (timeSpec.startsWith('毎月')) {
                    const pattern = timeSpec.match(/毎月(\d{1,2})日(\d{1,2}):(\d{1,2})/);
                    if (pattern) {
                        const [_, day, hour, minute] = pattern.map(Number);
                        
                        await dynamoService.saveReminder({
                            id: Date.now().toString(),
                            groupId: lineEvent.source.groupId!,
                            message,
                            scheduleTime: Date.now(), // ダミー値
                            createdBy: lineEvent.source.userId!,
                            isCompleted: false,
                            isRecurring: true,
                            recurringPattern: {
                                dayOfMonth: day,
                                hour,
                                minute
                            }
                        });

                        await lineClient.replyMessage(lineEvent.replyToken, {
                            type: 'text',
                            text: `毎月${day}日${hour}:${minute}にリマインドを設定しました: ${message}`
                        });
                    }
                } else {
                    const [_, timeStr, ...messageParts] = text.split(' ');
                    const message = messageParts.join(' ');
                    
                    const scheduleTime = new Date();
                    if (timeStr.endsWith('分後')) {
                        const minutes = parseInt(timeStr);
                        scheduleTime.setMinutes(scheduleTime.getMinutes() + minutes);
                    }

                    await dynamoService.saveReminder({
                        id: Date.now().toString(),
                        groupId: lineEvent.source.groupId!,
                        message,
                        scheduleTime: scheduleTime.getTime(),
                        createdBy: lineEvent.source.userId!,
                        isCompleted: false
                    });

                    await lineClient.replyMessage(lineEvent.replyToken, {
                        type: 'text',
                        text: `${timeStr}にリマインドを設定しました: ${message}`
                    });
                }
            }
        }
    }

    return { statusCode: 200, body: 'OK' };
};
